Multiprocessing

SendSlippi opened the pull request:
This is a big one. I initially wanted to separate out the multiprocessing from resolution/bitrate/widescreen, but they became so critical to determining how many games can be run at the same time that it made sense to expose them to the configuration file.

I decided that these should go into the config file instead of arguments. I think that a config file is a better approach should we add a GUI in the future. Plus, many arguments could be annoying to keep track of in the command line.

- Multiprocessing
  - When recording with a folder as the first argument, we can throw all the arguments into a multiprocessing pool and for record_single_slp to run against.
  - There was one bug that came out of this. Because all the arguments are thrown into a pool, I had to remove the ability to combine the files from a folder before finishing the entire pool. I documented this in the future work in the readme. But multiprocessing is so fast that it is worth it despite this hiccup.

- Resolution
  - Vertical resolution in dolphin is controlled in GFX.ini through a parameter called EFBScale. To alter it, I used a library called configparser which is designed to parse ini files. Each instance of the DolphinRunner class having it's own User folder made this very simple, so I'm glad that was added. The one thing I spent forever on was that the game-specific settings would take priority of the GFX.ini settings. So I had to patch the dolphin sys game settings folder so EFBScale can be set.

- Widescreen
  - Apply the 16:9 GFX aspect ratio and add the widescreen cheat to the list of enabled gecko codes.

- Bitrate
  - Apply bitrate to GFX

SendSlippi commented on slp2mp4/User/Config/Dolphin.ini:
```diff
@@ -31,7 +31,7 @@ DisableTooltips = False
 FullscreenResolution = Auto
 Fullscreen = False
-RenderToMain = False
+RenderToMain = True
```
It got annoying spawning more windows than necessary. This combines the video window with the main dolphin window.

SendSlippi replied:
Ok, I think I will make a PR after this to check if the system is windows and then apply a patch to set RenderToMain to true.

SendSlippi commented on slp2mp4/User/Config/GFX.ini:
```diff
@@ -69,14 +69,14 @@ ShowFrameTimes = False
 UseScalingFilter = True
 TextureScalingType = 0
-TextureScalingFactor = 2
+TextureScalingFactor = 1
```
Reduce some default effects to improve performance which can be better used by multiprocessing.

SendSlippi commented on slp2mp4/config.json:
```diff
@@ -2,6 +2,10 @@
     "melee_iso": "~/Games/SSBM-v1.02-NTSC.iso",
     "dolphin_dir": "~/dolphin/Slippi-FM-r18/playback",
     "ffmpeg": "/usr/bin/ffmpeg",
+    "resolution": "720p",
```
480p, 720p, 1080p, 1440p, 2160p

SendSlippi commented on slp2mp4/config.json:
```diff
@@ -2,6 +2,10 @@
     "dolphin_dir": "~/dolphin/Slippi-FM-r18/playback",
     "ffmpeg": "/usr/bin/ffmpeg",
+    "resolution": "720p",
+    "widescreen": true,
```
true, false

SendSlippi commented on slp2mp4/config.json:
```diff
@@ -2,6 +2,10 @@
     "ffmpeg": "/usr/bin/ffmpeg",
+    "resolution": "720p",
+    "widescreen": true,
+    "bitrateKbps": 16000,
```
Any nonzero value accepted by dolphin

SendSlippi commented on slp2mp4/config.json:
```diff
@@ -2,6 +2,10 @@
+    "resolution": "720p",
+    "widescreen": true,
+    "bitrateKbps": 16000,
+    "parallel_games": "recommended",
```
recommended, 1, 2, 3, 4, 5, etc.

SendSlippi commented on slp2mp4/dolphinrunner.py:
```diff
@@ -61,6 +62,29 @@ def count_frames_completed(self):
+        # Apply graphics settings
+        gfx_ini = os.path.join(self.user_dir, "Config", "GFX.ini")
+        gfx_ini_parser = configparser.ConfigParser()
+        gfx_ini_parser.optionxform = str
```
Need this to be case-sensitive

SendSlippi replied:
configparser uses a different type for it's elements I think? This sets them to be strings, which preserves the case-sensitivity of the file.

SendSlippi commented on slp2mp4/dolphinrunner.py:
```diff
@@ -61,6 +62,29 @@ def count_frames_completed(self):
+        if self.conf.widescreen:
+            gfx_ini_parser.set('Settings', 'AspectRatio', "6")
+            with open(os.path.join(self.user_dir, "GameSettings", "GALE01.ini"), "a") as game_settings_file:
+                game_settings_file.write("\n$Widescreen 16:9")
```
I decided to just append this to the file to enable the gecko code instead of using configparser because configparser didn't like '$'s.

SendSlippi commented on slp2mp4/installer.py:
```diff
@@ -32,6 +34,19 @@ def recursive_overwrite(src, dest, ignore=None):
+    gal_ini_parser.write(gal_ini_fp)
+    gal_ini_fp.close()
+
+
```
Have to patch the Sys/GameSettings/GAL.ini file here which sets the EFBScale.

NunoDasNeves replied:
See comment - this is broken on Linux

SendSlippi commented on slp2mp4/slp-to-mp4.py:
```diff
@@ -51,6 +51,19 @@ def is_game_too_short(num_frames, remove_short):
+    if conf.parallel_games == "recommended":
+        return psutil.cpu_count(logical=False)
+    else:
+        return int(conf.parallel_games)
```
Unfortunately we had to add psutil as a dependency because I could find no other way to get the physical cpu cores. multiprocess.cpu_count and os.cpu_count are both logical cores.

SendSlippi replied:
The recommended setting needs to estimate the number of parallel games that can be run on a CPU. This setting should hopefully prevent people from having to 'guess and check' the number of games that they can run. 

The biggest determining factor in running multiple dolphin instances is the number of CPU cores because it indicates the number of processes that can be run in parallel without the performance penalty of trying to run multiple processes on the same CPU. Logical cores can only simulate this but cannot get the same performance as the same number of physical cores when the physical cores are stressed.

I have an i7-8700k which has 6 physical cores and 12 logical cores. Running 12 games or more at a time is possible but the framerate drops significantly and is bottlenecked by CPU. I can run 7-9 games at a time (depending on resolution and bitrate) without seeing framerate dropping so physical cores are a better benchmark than logical cores. 

Many systems have the same number of physical and logical cores, in which case the recommendation would not matter. But for systems like mine where the difference between logical and physical cores is great, number of physical cores will always be a better predictor.

SendSlippi commented on slp2mp4/slp-to-mp4.py:
```diff
@@ -145,7 +161,13 @@ def record_folder_slp(slp_folder, conf):
+    num_processes = get_num_processes(conf)
+
+    pool = multiprocessing.Pool(processes=num_processes)
+    pool.starmap(record_file_slp, args)
```
starmap takes a function as the first argument and a list of tuples for the second.

NunoDasNeves commented on slp2mp4/installer.py:
```diff
@@ -32,6 +34,19 @@ def recursive_overwrite(src, dest, ignore=None):
+
+def patch_dolphin_sys_game_settings():
+    # Remove efb_scale field. This allows selection of resolution options from GFX.ini.
+    gal_ini = os.path.join(FM_WIN_FOLDER, "Sys", "GameSettings", "GAL.ini")
```
This breaks Linux compatibility because FM_WIN_FOLDER is hardcoded here :'(
I think the correct solution here is to patch GAL.ini on the fly, using whatever dolphin directory is specified in config.json.
Ideally, it patches it at the start of execution, and then un-patches it after, so it doesn't mess up people's playback settings...but this could still go go wrong... hmm it's a bit annoying tbh
Maybe copying the whole playback directory into slp2mp4, and running dolphin from there is better.

SendSlippi replied:
Aw I missed that I had selected that hardcoded windows path instead. Thanks for catching that. I'll switch it to patch from the config entry.

The most annoying part is that if you control-C or otherwise exit the python script we don't have a mechanism in place right now to catch it an unpatch this to exit gracefully (though we should at some point). This is why I had to create that 'clean()' function to run at startup; exiting early would not remove the temporary files.

For this specific patch, I'm not too concerned about it messing up people's playback settings. It just means it will be taking the EFBScale configuration from GFX.ini by default.

NunoDasNeves commented on slp2mp4/slp-to-mp4.py:
```diff
@@ -129,12 +142,15 @@ def record_folder_slp(slp_folder, conf):
             if conf.combine:
-                combine(conf)
+                #combine(conf)
+                # TODO: Find a way to fix combining folders while still multiprocessing
```
Easiest way to do this is probably multiprocess one directory at a time

NunoDasNeves replied:
Can be saved for another PR though

SendSlippi replied:
That's the easiest way to do it, but it impacts throughput because you
- Can only record up to the number of games in the directory instead of what can be supported by the system. E.g. if my computer can run 4 games at a time but we have two folders of 2 .slp files, what we have now can record them ~twice as fast.
- You have to wait for the last game to end before starting to record the next directory. The last game in the recording process always slows throughput because only one game is recording at that time.

For fastest throughput, we would have something that detects when all the valid games are in the output folder and then combine them, but it has to happen while the multiprocessing is running.

I agree that it should wait until another PR.

NunoDasNeves commented on slp2mp4/slp-to-mp4.py:
```diff
@@ -193,4 +215,6 @@ def main():
 
 if __name__ == '__main__':
     installDependencies()
+    patch_dolphin_sys_game_settings()
```
This is a style/preference thing, but I think main() should be the only thing here; these other functions should just be be called from within main itself.

My thinking is that the whole `if __name__ == '__main__':` thing is just boilerplate that has to be there; I prefer not to put any actual logic there other than the call to main

SendSlippi replied:
Makes sense

SendSlippi commented on slp2mp4/dolphinrunner.py:
```diff
@@ -146,7 +149,8 @@ def run(self, slp_file, num_frames):
                 if time.perf_counter() - start_timer > MAX_WAIT_SECONDS:
-                    raise RuntimeError("Timed out waiting for render")
+                    print("WARNING: Timed out waiting for render")
+                    break
```
I noticed that on games that went to time sometimes this condition was met and it would stop all other games from being recorded. So I bumped up the time and made it so it only prints a warning, then breaks and kills dolphin like normal.

SendSlippi commented:
I addressed comments so hopefully linux shouldn't see those problems now. I also made it so games that exceed the time don't raise an exception and stop the entire recording process. It will just print a warning, kill dolphin and continue.
